packages/runtime-core/src/componentSlots.ts

```diff
@@ -87,17 +88,18 @@ const normalizeSlotValue = (value: unknown): VNode[] =>
 const normalizeSlot = (
   key: string,
   rawSlot: Function,
-  ctx: ComponentInternalInstance | null | undefined,
+  rawSlots: RawSlots,
 ): Slot => {
   if ((rawSlot as any)._n) {
     // already normalized - #5353
     return rawSlot as Slot
   }
+  const ctx = rawSlots._ctx
   const normalized = withCtx((...args: any[]) => {
     if (
       __DEV__ &&
       currentInstance &&
-      (!ctx || ctx.root === currentInstance.root)
+      (!hasOwn(rawSlots, '_ctx') || (ctx && ctx.root === currentInstance.root))
```

Comment:
The condition will cause the following warning to never trigger. Removing the condition `|| (ctx && ctx.root === currentInstance.root)` will also prevent the warning from triggering in https://github.com/vuejs/core/blob/41e10dce085768a95f83cda175aef0c6cc42b2dd/packages/runtime-core/__tests__/componentSlots.spec.ts#L324. I think a test case needs to be added to verify the warning will trigger.

Reply:
Is there a slot case that triggers warn?

Reply:
https://github.com/vuejs/core/blob/cb34b28a4a9bf868be4785b001c526163eda342e/packages/runtime-core/src/vnode.ts#L853

if the slot value is a function the `rawSlots` always has the `_ctx` key.

so all test cases pass using the following conditions
```ts
if (
  __DEV__ &&
  currentInstance &&
  (ctx && ctx.root === currentInstance.root)
```

Reply:
Oh, I see, then the following [playground](https://play.vuejs.org/#eNp9VF1v0zAU/SvGD2sqhRQ0eOnaIhiVAGkDAYIXS1OW3LYZjm35I3Sq8t+5dlonYVMfWtX3nHvvuV890PdKZY0DOqcLU+hKWWLAOrVioqqV1JYciIYNaclGy5pMkDoZQFsQoHMLN7LM+YnEaDarvYFRJpgopDCW1GZLlj5UMvkEnEvyW2pevphMrzxnFCeZMrGYdWpQBz4s1IojjC9CFrvXq8MhBGzbxQxfwVoJ5SxpXmJm4EtGEWeUzBBczAb+NKXWoKRNtc0ejBRY+MH7M1rIWlUc9FdlK5TM6JwExGM5Kv77JdisdpCe7MUOij/P2B/M3tsY/abBgG6A0YjZXG/BdvD6xy3s8XcEUb3jyD4DfgcjufMaO9oHJ0qUPeAFtZ/DjCqx/WnWewvCnIryQj2zDXxGcaTXZ0rv5V5mb4IfEy12MUwYW4gNHG5KuksLDdjrX7c4iLSETSVCfClA2FSDKEGP10njjPED+xClW5fxYi1JMiXLFYpCHjlSwNtLWbgaA2dd0jUH/0omZdX41RrQnbiRTtj/YxEU7SUlwnGeYtCTF8HfmcoRtb6Sd5mGWjZwvat4mfS0dpSkPpviY5VzuR0niQXcy/Ixy5VC6vM5hnnw2+bYWO1bMG5xErOGQ07uUhyM4dIa0k57SV6UdVqQjROFn7rXPEAjHlwzTJI7bsnFxdiQ9JX0vQi71Skfz6BrAGreJbGCtGt8TB0DP5XadXWwXslQL6PhL8DfwcA6KokQ+6hgfrzbe4lHOoTb/tGXdTTGSsKIY9lHZTFLQKOA486dmhF82is8nrsGtD9IPJ7L7G32irb/ANpL0GI=) is also not as expected 
I said it was a bit weird

Reply:
I'm sorry. I still don't understand the problem here
[playground](https://play.vuejs.org/#eNq1Ul1PgzAU/StNHwQSApnzaW5L1OxBH9Sob9YYAt1glpb0Y5Is/Hdvy0C3jPmiCQ/ce07bc889W3xVVdHGUDzBU5XKotJzwouyElKjG1FWaClFibwotoUlepd7+GiPMDpgbFGOmh1jhxBOa4dldJkYBhzCEVJUm8oP2gIhCaXkCBqzOcp9Lys2Xoi4YSxEr7lvn+rKnqOqhAPJG52PvSAIoeUU/cJ7C+yLDeHwTePeARxirVLBl8UqWivBwR4njeAULi0YlQ+VLgRXBE860QQnjInPO9fT0tCw66c5TT+O9Neqtj2CHyVVVG4owT2mE7miuoUXz/e0hv8eLEVmGLBPgE9UCWasxpZ2bXgGsn/wnNpbt6iCr17UotaUq24oK9RZ4/gEw/qsn0Ojf8sdRxedpeBiF5tj+TqSjuFsvIfAV0xohZo+J3EMO1IaaXAAzVo46g6fne03fLfq0+E6ODAYjj7s/zPXH09lLxoYpfkCkD5NBw==)

Reply:
I've added a few test cases, are they correct?

Reply:
All test cases are correct.

Reply:
cool!!!